src/meta/src/hummock/manager/mod.rs

```diff
@@ -1109,136 +1172,222 @@ where
         Ok(true)
     }
 
-    fn sync_group(
-        old_version_groups: Vec<CompactionGroupId>,
-        compaction_groups: HashMap<CompactionGroupId, CompactionGroup>,
-        new_version_delta: &mut BTreeMapEntryTransaction<'_, HummockVersionId, HummockVersionDelta>,
-        new_hummock_version: &mut HummockVersion,
-    ) {
+    async fn sync_group<'a>(
+        &'a self,
+        compaction: &'a mut Compaction,
+        versioning: &'a mut Versioning,
+        compaction_groups: &mut HashMap<CompactionGroupId, CompactionGroup>,
+    ) -> Result<(u64, HummockVersionDelta, HummockVersion)> {
         // We need 2 steps to sync groups:
         // Insert new groups that are not in current `HummockVersion`;
         // Delete old groups that still remain in current `HummockVersion`.
+        let mut old_version = versioning.current_version.clone();
+        let old_version_groups = old_version
+            .get_levels()
+            .iter()
+            .map(|(group_id, _)| *group_id)
+            .collect_vec();
+        let mut new_version_id = old_version.id + 1;
+        let mut new_version_delta = BTreeMapEntryTransaction::new_insert(
+            &mut versioning.hummock_version_deltas,
+            new_version_id,
+            HummockVersionDelta {
+                prev_id: old_version.id,
+                safe_epoch: old_version.safe_epoch,
+                trivial_move: false,
+                ..Default::default()
+            },
+        );
+
+        let mut new_hummock_version = old_version;
+        new_version_delta.id = new_version_id;
+        new_hummock_version.id = new_version_id;
+
+        let mut need_independent_commit = false;
+        for group_id in &old_version_groups {
+            if !compaction_groups.contains_key(group_id) {
+                need_independent_commit = true;
+            }
+        }
+        for group_id in compaction_groups.keys() {
+            if !new_hummock_version.levels.contains_key(group_id) {
+                need_independent_commit = true;
+            }
+        }
+        if !need_independent_commit {
+            return Ok((
+                new_version_delta.key,
+                new_version_delta.new_value,
+                new_hummock_version,
+            ));
+        }
```

Comment:
> Also, why do we need to geneerate a new version here if there is no group changed?

If Some SSTs are broken, we need to loop
sync_group, commit_epoch and then rollback both cause SST broken
sync_group, commit_epoch and then rollback both cause SST still broken
sync_group, commit_epoch and then rollback both cause SST broken again
To prevent this, we commit sync_group first. Thus we loop
commit_epoch and then rollback commit_epoch only

Reply:
> Can be simplified with `all`:
> 
> ```rust
> let group_unchanged = old_version_groups.iter().all(|group_id| compaction_groups.contains_key(group_id))
>             && compaction_groups.keys().all(|group_id| new_hummock_version.levels.contains_key(group_id));
> if group_unchanged {
>     return Ok((
>         new_version_delta.key,
>         new_version_delta.new_value,
>         new_hummock_version,
>     ));
> }
> ```

fixed

Reply:
> Also, why do we need to geneerate a new version here if there is no group changed?

fixed